### PR TITLE
Allow passing torch schedulers to torch_minimize and general CEEM hp scheduler

### PR DESCRIPTION
**Pull request by rejuvyesh**
Currently would not support `ReduceLROnPlateau` which has different stepping protocol.
Also only supporting schedulers available in `torch.optim.lr_scheduler`.

### CODE REVIEW

**Comment by kunalmenda** [comment]
An issue here is that the optimizer resets every learning step. Ideally the scheduling happens every CEEM epoch. Maybe we can make the scheduled state persist between epochs?

**Comment by rejuvyesh** [comment]
> An issue here is that the optimizer resets every learning step. Ideally the scheduling happens every CEEM epoch. Maybe we can make the scheduled state persist between epochs?

torch schedulers wrap optimizers, so we'll have to store the `scheduler.state_dict()` somewhere and load that again maybe?

**Comment by rejuvyesh** [comment]
Otherwise we need our own schedulers that work in `CEEM.step` independent of torch schedulers.

**Comment by kunalmenda** [comment]
How about this: we dont really need the lr to anneal over the 100-1000 epochs in a single learning epoch, but just across them. We could potentially pass an optional `lr_scheduler` to the learner every learning epoch and have the learner just call:
```
lr_scheduler.step()
lr = base_lr * lr_scheduler.lr_scale
```
Or something?

We can just let `lr_scheduler` be a class instance with `step()` and `lr_scale`.

**Comment by rejuvyesh** [comment]
Why not modify the learning rate in `opt_kwargs` in `CEEM.step` instead using our own lr_scheduler? No need to pass it to the learner at all?

**Comment by kunalmenda** [comment]
Yeah thats another option too. So you're thinking we pass the scheduler just to `ceem.train`?

**Comment by rejuvyesh** [comment]
@kunalmenda How about now? This allows a general scheduler for all hyperparameters in the optimizer and should work for non torch things as well.

File: ceem/ceem.py [hunk 1]
```diff
@@ -118,6 +118,19 @@ def step(self, xs, sys, smooth_solver_kwargs=None, learner_criterion_kwargs=None
...
+        if hp_schedulers is not None:
+            if isinstance(hp_schedulers, dict):
+                hp_schedulers = [hp_schedulers for _ in range(len(self._learning_criteria))]
+            else:
+                assert len(hp_schedulers) == len(self._learning_criteria)
+
+        if hp_schedulers is not None:
+            for i in range(len(self._learning_criteria)):
+                for hp, scheduler in hp_schedulers[i].items():
+                    newhp = scheduler.step()
```
Comment (kunalmenda) [inline ceem/ceem.py:125]: `step()` will be called multiple times in one epoch if line 123 is called right?

Comment (rejuvyesh) [reply ceem/ceem.py:125]: Hmm. I should deepcopy it instead. Good catch!

Comment (rejuvyesh) [reply ceem/ceem.py:125]: But the setup should happen in train, you are right.

Comment (rejuvyesh) [reply ceem/ceem.py:125]: For the same criteria there can be multiple schedulers for the different hyper parameters, so this is fine?

Comment (kunalmenda) [reply ceem/ceem.py:125]: The issue I see is:
1. if you deep copy, then when you `step`, the original object from which you deepcopied won't stepped. 
2. if you have multiple criteria (eg dynamics, observations) but want to use the same scheduler for them, then (if you didn't deep copy) you would call `step` twice.

I filed a PR fix to this, maybe see if that works?

Comment (kunalmenda) [reply ceem/ceem.py:125]: I see what you did. Theres just one issue with the way you've set up `LambdaScheduler` I think. I'll point it out.

File: ceem/utils.py [hunk 2]
```diff
@@ -99,3 +99,27 @@ def get_grad_norm(params):
...
+class LambdaScheduler(Scheduler):
+    def __init__(self, init_val, val_lambda, last_epoch=-1):
+        self.val_lambda = val_lambda
+        super().__init__(init_val, last_epoch)
+
+    def step(self, epoch=None):
+        if epoch is not None:
+            self.last_epoch = epoch
+
+        self._last_val = self.init_val * self.val_lambda(self.last_epoch)
```
Comment (kunalmenda) [inline ceem/utils.py:125]: Since we're not calling `step` with the epoch, should we be incrementing `last_epoch` every time step is called? An even safer approach would be to make the scheduler stateless and just have it be a function mapping `epoch -> param`.

**Comment by rejuvyesh** [comment]
@kunalmenda Let me know if you want me to remove the torch scheduler. But the way I see it, it's not a bad feature to have anyway?

**Comment by kunalmenda** [comment]
Ohh the schedulers in `utils` are for within `learner`, not the `hp_scheduler`?

**Review by kunalmenda — COMMENTED** [review]
> Ohh the schedulers in `utils` are for within `learner`, not the `hp_scheduler`?

Wait no, `torch_minimize` expects a torch scheduler. Then I think the `Scheduler` in `utils` should be incrementing its own epoch and not expected it to be passed. Or we can pass it the epoch from `step` but then we'll also need `train` to pass the epoch to `step`.

File: ceem/utils.py (hunk 2 above)
Comment (rejuvyesh) [reply ceem/utils.py:125]: Oh nooo, I forgot to do that. Good catch!

Comment (rejuvyesh) [reply ceem/utils.py:125]: So yeah, we can already pass in the epoch and make it stateless if you prefer that. Let me know.

**Comment by kunalmenda** [comment]
I think it looks good now! I'll merge it.
